CAPTCHA

### Description

- Currently only a static image is displayed in the place of captcha
- Find ways to generate real captcha and match it with user input.
- Discover React libraries to implement captcha Feature.
- The randomly generated captcha image being similar to the current static image is preferred style-wise
- However any new modifications in the generated image is allowed for consideration. 

### Screenshots

- Captcha Image

![Screenshot_2022-10-04-11-05-03-91_40deb401b9ffe8e1df2f1cc5ba480b12](https://user-images.githubusercontent.com/75939390/193742769-856f17db-0ea0-4053-b919-55443c9f3958.jpg)


### Additional information

## Make sure to read the [CONTRIBUTING](https://github.com/warmachine028/uemcrp/blob/main/CONTRIBUTING.md) and [SETUP](https://github.com/warmachine028/uemcrp/tree/main/rules) docs before proceeding
## Happy contributing. 💐
Star my other Repositories [here](https://github.com/warmachine028)

> Can you do `npm i --force` that might resolve?
 
 Sorry, I'm still I'm getting these errors on npm start.



Here's the log:
```
Failed to compile.

Module not found: Error: Can't resolve '@progress/kendo-drawing' in 'D:\CODING\contributions\uemcrp\client\node_modules\@progress\kendo-react-pdf\dist\es'
ERROR in ./node_modules/@progress/kendo-react-pdf/dist/es/PDFExport.js 28:0-61
Module not found: Error: Can't resolve '@progress/kendo-drawing' in 'D:\CODING\contributions\uemcrp\client\node_modules\@progress\kendo-react-pdf\dist\es'

ERROR in ./node_modules/@progress/kendo-react-pdf/dist/es/PDFExport.js 32:0-55
Module not found: Error: Can't resolve '@progress/kendo-licensing' in 'D:\CODING\contributions\uemcrp\client\node_modules\@progress\kendo-react-pdf\dist\es'

ERROR in ./node_modules/@progress/kendo-react-pdf/dist/es/savePDF.js 1:0-61
Module not found: Error: Can't resolve '@progress/kendo-drawing' in 'D:\CODING\contributions\uemcrp\client\node_modules\@progress\kendo-react-pdf\dist\es'

webpack compiled with 3 `errors`

```

--- a/client/package.json
+++ b/client/package.json
@@ -14,6 +14,7 @@
 		"@testing-library/jest-dom": "^5.16.5",
 		"@testing-library/react": "^13.3.0",
 		"@testing-library/user-event": "^13.5.0",
+		"buffer": "^6.0.3",
 		"moment": "^2.29.4",
 		"react": "^18.2.0",
 		"react-calendar": "^3.7.0",
@@ -25,7 +26,7 @@
 		"web-vitals": "^2.1.4"
 	},
 	"browser": {
-    "buffer": false
+		"buffer": false
 	},
 	"scripts": {
 		"start": "react-scripts start",

--- a/client/package.json
+++ b/client/package.json
@@ -24,6 +24,9 @@
 		"react-swipeable-views-react-18-fix": "^0.14.1",
 		"web-vitals": "^2.1.4"
 	},
+	"browser": {
+    "buffer": false
+	},
 	"scripts": {
 		"start": "react-scripts start",
 		"build": "react-scripts build",

--- a/client/package.json
+++ b/client/package.json
@@ -8,6 +8,8 @@
 		"@mui/icons-material": "^5.10.2",
 		"@mui/material": "^5.10.2",
 		"@mui/x-date-pickers": "^5.0.0",
+		"@progress/kendo-drawing": "^1.17.1",
+		"@progress/kendo-licensing": "^1.2.2",
 		"@progress/kendo-react-pdf": "^5.5.0",
 		"@testing-library/jest-dom": "^5.16.5",
 		"@testing-library/react": "^13.3.0",
@@ -18,6 +20,7 @@
 		"react-dom": "^18.2.0",
 		"react-router-dom": "^6.3.0",
 		"react-scripts": "5.0.1",
+		"react-simple-captcha": "^9.0.1",
 		"react-swipeable-views-react-18-fix": "^0.14.1",
 		"web-vitals": "^2.1.4"
 	},

--- a/client/src/components/Auth/index.jsx
+++ b/client/src/components/Auth/index.jsx
@@ -1,111 +1,189 @@
-import { Grid, Button, Select, MenuItem, InputLabel, FormControl, Checkbox } from "@mui/material"
-import Input from "../Input"
-import { LockOutlined } from "@mui/icons-material"
-import { Link, useNavigate } from "react-router-dom"
-import { useState } from "react"
-import { Root, classes, BpCheckedIcon, BpIcon } from "./styles"
-import Form from "../Form"
+import {
+  Grid,
+  Button,
+  Select,
+  MenuItem,
+  InputLabel,
+  FormControl,
+  Checkbox,
+} from "@mui/material";
+import Input from "../Input";
+import { LockOutlined } from "@mui/icons-material";
+import { Link, useNavigate } from "react-router-dom";
+import { useEffect, useState } from "react";
+import { Root, classes, BpCheckedIcon, BpIcon } from "./styles";
+import Form from "../Form";
 
 // * Dynamic Objects
 // * These objects are to be fetched OR created dynamically
-import captcha from "../../images/captcha.png"
-import { students } from "../../data/users"
+import captcha from "../../images/captcha.png";
+import { students } from "../../data/users";
+
+import {
+  loadCaptchaEnginge,
+  LoadCanvasTemplateNoReload,
+  validateCaptcha,
+} from "react-simple-captcha";
 
 const initialState = {
-	enrollment: "", //
-	password: "",
-	captcha: "",
-	role: "student",
-	remember: false,
-	captchaImg: captcha,
-}
+  enrollment: "", //
+  password: "",
+  captcha: "",
+  role: "student",
+  remember: false,
+  captchaImg: captcha,
+};
 
 const Auth = ({ setUser }) => {
-	const [formData, setFormData] = useState(initialState)
-	const [error, setError] = useState({ enrollment: false, password: false, captcha: false })
-	const [showPassword, setShowPassword] = useState(false)
-	const handleChange = event => setFormData({ ...formData, [event.target.name]: event.target.value })
-	const handleShowPassword = () => setShowPassword(prevShowPassword => !prevShowPassword)
-	const handleRemember = () => setFormData({ ...formData, remember: !formData.remember })
-	const navigate = useNavigate()
-	const user = students[0]
-	const handleSubmit = event => {
-		event.preventDefault()
-		if (validate()) {
-			setUser(user)
-			navigate("/")
-		}
-	}
-	const validate = () => {
-		const errors = {
-			enrollment: !formData.enrollment.match(/^[0-9]*$/) || formData.enrollment.length < 13,
-			password: formData.password.length < 6,
-			captcha: formData.captcha !== "G2KS",
-		}
-		setError(errors)
-		return !(errors.enrollment || errors.password || errors.captcha)
-	}
-	return (
-		<Root className={classes.root}>
-			<Form icon={<LockOutlined />} title="SIGN IN" submit={handleSubmit}>
-				<Input error={error.enrollment} name="enrollment" value={formData.enrollment} label="User Name" onChange={handleChange} />
-				<Input
-					name="password"
-					error={error.password}
-					value={formData.password}
-					label="Password"
-					onChange={handleChange}
-					type={showPassword ? "text" : "password"}
-					handleShowPassword={handleShowPassword}
-				/>
-				<img src={formData.captchaImg} alt="captcha" className={classes.captchaImg} />
+  const [formData, setFormData] = useState(initialState);
+  const [error, setError] = useState({
+    enrollment: false,
+    password: false,
+    captcha: false,
+  });
+  const [showPassword, setShowPassword] = useState(false);
+  const handleChange = (event) => {
+    setFormData({ ...formData, [event.target.name]: event.target.value });
+  };
+  const handleShowPassword = () =>
+    setShowPassword((prevShowPassword) => !prevShowPassword);
+  const handleRemember = () =>
+    setFormData({ ...formData, remember: !formData.remember });
+  const navigate = useNavigate();
+  const user = students[0];
 
-				<Input error={error.captcha} value={formData.captcha} name="captcha" label="Captcha" onChange={handleChange} />
-				<FormControl fullWidth style={{ padding: "5px 0px" }}>
-					<InputLabel id="role-label" style={{ top: 5 }}>
-						Role
-					</InputLabel>
-					<Select fullWidth label="Role" labelId="role-label" value={formData.role} onChange={handleChange} name="role">
-						<MenuItem name="role" value="teacher">
-							Staff & Faculty
-						</MenuItem>
-						<MenuItem name="role" value="student">
-							Student
-						</MenuItem>
-					</Select>
-				</FormControl>
+  const loadCaptcha = () => {
+    // captcha background
+    const angle = (90 * Math.PI) / 180;
+    const x2 = 50 * Math.cos(angle);
+    const y2 = 50 * Math.sin(angle);
 
-				<Button type="submit" fullWidth className={classes.submit} style={{ color: "white" }}>
-					SIGN IN
-				</Button>
-				<Grid container justifyContent="space-between">
-					<Button style={{ paddingLeft: 0 }} onClick={handleRemember} value={formData.remember} disableElevation disableRipple>
-						<BpCheckbox checked={formData.remember} />
-						Remember Me
-					</Button>
-					<Button component={Link} to="/auth">
-						Forgot Password
-					</Button>
-				</Grid>
-			</Form>
-		</Root>
-	)
-}
+    const bg = document
+      .createElement("canvas")
+      .getContext("2d")
+      .createLinearGradient(0, 0, x2, y2);
 
-const BpCheckbox = props => {
-	return (
-		<Checkbox
-			sx={{
-				"&:hover": { bgcolor: "transparent" },
-			}}
-			checked={props.checked}
-			disableRipple
-			color="default"
-			checkedIcon={<BpCheckedIcon />}
-			icon={<BpIcon />}
-			inputProps={{ "aria-label": "Checkbox demo" }}
-			{...props}
-		/>
-	)
-}
-export default Auth
+    bg.addColorStop(0, "black");
+    bg.addColorStop(1, "red");
+
+    loadCaptchaEnginge(4, bg, "#f19818");
+  };
+
+  useEffect(() => {
+    loadCaptcha();
+  }, []);
+
+  const handleSubmit = (event) => {
+    event.preventDefault();
+    if (validate()) {
+      setUser(user);
+      navigate("/");
+    }
+  };
+  const validate = () => {
+    const errors = {
+      enrollment:
+        !formData.enrollment.match(/^[0-9]*$/) ||
+        formData.enrollment.length < 13,
+      password: formData.password.length < 6,
+      captcha: !validateCaptcha(formData.captcha),
+    };
+    setError(errors);
+    return !(errors.enrollment || errors.password || errors.captcha);
+  };
+  return (
+    <Root className={classes.root}>
+      <Form icon={<LockOutlined />} title="SIGN IN" submit={handleSubmit}>
+        <Input
+          error={error.enrollment}
+          name="enrollment"
+          value={formData.enrollment}
+          label="User Name"
+          onChange={handleChange}
+        />
+        <Input
+          name="password"
+          error={error.password}
+          value={formData.password}
+          label="Password"
+          onChange={handleChange}
+          type={showPassword ? "text" : "password"}
+          handleShowPassword={handleShowPassword}
+        />
+        {/* <img src={formData.captchaImg} alt="captcha" className={classes.captchaImg} /> */}
+
+        <LoadCanvasTemplateNoReload />
+
+        <Input
+          error={error.captcha}
+          value={formData.captcha}
+          name="captcha"
+          label="Captcha"
+          onChange={handleChange}
+        />
+        <FormControl fullWidth style={{ padding: "5px 0px" }}>
+          <InputLabel id="role-label" style={{ top: 5 }}>
+            Role
+          </InputLabel>
+          <Select
+            fullWidth
+            label="Role"
+            labelId="role-label"
+            value={formData.role}
+            onChange={handleChange}
+            name="role"
+          >
+            <MenuItem name="role" value="teacher">
+              Staff & Faculty
+            </MenuItem>
+            <MenuItem name="role" value="student">
+              Student
+            </MenuItem>
+          </Select>
+        </FormControl>
+
+        <Button
+          type="submit"
+          fullWidth
+          className={classes.submit}
+          style={{ color: "white" }}
+        >
+          SIGN IN
+        </Button>
+        <Grid container justifyContent="space-between">
+          <Button
+            style={{ paddingLeft: 0 }}
+            onClick={handleRemember}
+            value={formData.remember}
+            disableElevation
+            disableRipple
+          >
+            <BpCheckbox checked={formData.remember} />
+            Remember Me
+          </Button>
+          <Button component={Link} to="/auth">
+            Forgot Password
+          </Button>
+        </Grid>
+      </Form>
+    </Root>
+  );
+};
+
+const BpCheckbox = (props) => {
+  return (
+    <Checkbox
+      sx={{
+        "&:hover": { bgcolor: "transparent" },
+      }}
+      checked={props.checked}
+      disableRipple
+      color="default"
+      checkedIcon={<BpCheckedIcon />}
+      icon={<BpIcon />}
+      inputProps={{ "aria-label": "Checkbox demo" }}
+      {...props}
+    />
+  );
+};
+export default Auth;

--- a/client/src/components/Auth/styles.js
+++ b/client/src/components/Auth/styles.js
@@ -1,69 +1,76 @@
-import { styled } from "@mui/material/styles"
+import { styled } from "@mui/material/styles";
 
-const PREFIX = "Auth"
+const PREFIX = "Auth";
 export const classes = {
-	root: `${PREFIX}-root`,
-	captchaImg: `${PREFIX}-captchaImg`,
-	submit: `${PREFIX}-submit`,
-}
+  root: `${PREFIX}-root`,
+  captchaImg: `${PREFIX}-captchaImg`,
+  submit: `${PREFIX}-submit`,
+};
 
 export const BpIcon = styled("span")(({ theme }) => ({
-	borderRadius: 3,
-	width: 16,
-	height: 16,
-	boxShadow: theme.palette.mode === "dark" ? "0 0 0 1px rgb(16 22 26 / 40%)" : "inset 0 0 0 1px rgba(16,22,26,.2), inset 0 -1px 0 rgba(16,22,26,.1)",
-	backgroundColor: theme.palette.mode === "dark" ? "#394b59" : "#f5f8fa",
-	backgroundImage: theme.palette.mode === "dark" ? "linear-gradient(180deg,hsla(0,0%,100%,.05),hsla(0,0%,100%,0))" : "linear-gradient(180deg,hsla(0,0%,100%,.8),hsla(0,0%,100%,0))",
-	".Mui-focusVisible &": {
-		outline: "2px auto black",
-		outlineOffset: 2,
-	},
-	"input:hover ~ &": {
-		backgroundColor: theme.palette.mode === "dark" ? "#30404d" : "#ebf1f5",
-	},
-	"input:disabled ~ &": {
-		boxShadow: "none",
-		background: theme.palette.mode === "dark" ? "rgba(57,75,89,.5)" : "rgba(206,217,224,.5)",
-	},
-}))
+  borderRadius: 3,
+  width: 16,
+  height: 16,
+  boxShadow:
+    theme.palette.mode === "dark"
+      ? "0 0 0 1px rgb(16 22 26 / 40%)"
+      : "inset 0 0 0 1px rgba(16,22,26,.2), inset 0 -1px 0 rgba(16,22,26,.1)",
+  backgroundColor: theme.palette.mode === "dark" ? "#394b59" : "#f5f8fa",
+  backgroundImage:
+    theme.palette.mode === "dark"
+      ? "linear-gradient(180deg,hsla(0,0%,100%,.05),hsla(0,0%,100%,0))"
+      : "linear-gradient(180deg,hsla(0,0%,100%,.8),hsla(0,0%,100%,0))",
+  ".Mui-focusVisible &": {
+    outline: "2px auto black",
+    outlineOffset: 2,
+  },
+  "input:hover ~ &": {
+    backgroundColor: theme.palette.mode === "dark" ? "#30404d" : "#ebf1f5",
+  },
+  "input:disabled ~ &": {
+    boxShadow: "none",
+    background:
+      theme.palette.mode === "dark"
+        ? "rgba(57,75,89,.5)"
+        : "rgba(206,217,224,.5)",
+  },
+}));
 
 export const BpCheckedIcon = styled(BpIcon)({
-	backgroundColor: "black",
-	backgroundImage: "linear-gradient(180deg,hsla(0,0%,100%,.1),hsla(0,0%,100%,0))",
-	"&:before": {
-		display: "block",
-		width: 16,
-		height: 16,
-		backgroundImage:
-			"url(\"data:image/svg+xml;charset=utf-8,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16'%3E%3Cpath" +
-			" fill-rule='evenodd' clip-rule='evenodd' d='M12 5c-.28 0-.53.11-.71.29L7 9.59l-2.29-2.3a1.003 " +
-			"1.003 0 00-1.42 1.42l3 3c.18.18.43.29.71.29s.53-.11.71-.29l5-5A1.003 1.003 0 0012 5z' fill='%23fff'/%3E%3C/svg%3E\")",
-		content: '""',
-	},
-	"input:hover ~ &": {
-		backgroundColor: "gray",
-	},
-})
+  backgroundColor: "black",
+  backgroundImage:
+    "linear-gradient(180deg,hsla(0,0%,100%,.1),hsla(0,0%,100%,0))",
+  "&:before": {
+    display: "block",
+    width: 16,
+    height: 16,
+    backgroundImage:
+      "url(\"data:image/svg+xml;charset=utf-8,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16'%3E%3Cpath" +
+      " fill-rule='evenodd' clip-rule='evenodd' d='M12 5c-.28 0-.53.11-.71.29L7 9.59l-2.29-2.3a1.003 " +
+      "1.003 0 00-1.42 1.42l3 3c.18.18.43.29.71.29s.53-.11.71-.29l5-5A1.003 1.003 0 0012 5z' fill='%23fff'/%3E%3C/svg%3E\")",
+    content: '""',
+  },
+  "input:hover ~ &": {
+    backgroundColor: "gray",
+  },
+});
 
 export const Root = styled("div")(({ theme }) => ({
-	[`&.${classes.root}`]: {
-		"& .MuiButtonBase-root": {
-			color: "black",
-		},
-	},
-	[`& .${classes.captchaImg}`]: {
-		paddingTop: 3,
-		width: "100%",
-		height: 50,
-		objectFit: "cover",
-		borderRadius: 5,
-		objectPosition: "0 -37px",
-	},
-	[`& .${classes.submit}`]: {
-		margin: "10px 0",
-		color: "white",
-		backgroundColor: "#393464",
-		"&:hover": { backgroundColor: "#564f95" },
-	},
-
-}))
+  [`&.${classes.root}`]: {
+    "& .MuiButtonBase-root": {
+      color: "black",
+    },
+  },
+  [`& .${classes.captchaImg}, & canvas`]: {
+    paddingTop: 3,
+    width: "100%",
+    height: 50,
+    borderRadius: 5,
+  },
+  [`& .${classes.submit}`]: {
+    margin: "10px 0",
+    color: "white",
+    backgroundColor: "#393464",
+    "&:hover": { backgroundColor: "#564f95" },
+  },
+}));

